<a href="https://colab.research.google.com/github/silvalugo/python-for-ai/blob/main/02_data_analysis_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src='images/Practicum_AI_Logo.white_outline.svg' width=250 alt='Practicum AI logo'> <img src='https://github.com/PracticumAI/practicumai.github.io/blob/main/images/icons/practicumai_python.png?raw=true' align='right' width=50>
***

# Data Analysis with pandas
In this module, you will learn about one of the most popular and valuable Python libraries for data analysis and manipulation: ***pandas***!

<img src='images/pandas_logo.png' alt='The pandas logo' align='center' width=200>

### After this module, students will be able to:
* Explain the various components of a pandas DateFrame.
* Use `read_csv()` to create new DateFrames from data.
* Subset data with pandas index and DateFrame management functions.
* Use basic statistical functions to summarize a dataset.
* Use matplotlib functions to generate clear graphs and plots.


## <img src='images/get_started_icon.svg' alt="Let's get started header" width=40 align=center> Let's Get Started!

As we learned in the last module, we must **import** the pandas library before we can use it.
* We'll also import the scientific computing library NumPy for illustration purposes.
* NumPy and pandas are usually imported with abbreviated aliases, `np` and `pd`.

In [ ]:
import numpy as np
import pandas as pd

>## <img src='images/alert_icon.svg' alt="Alert icon" width=40 align=center> Did you get an error here?
>
>If you got an error running the cell above, it is probably because the environment you are working in doesn't have `numpy` or `pandas` installed.
>
>### For Google Colab
>
>On Google Colab, this shouldn't be an issue (for these libraries), and if you do run into this in the future, it is generally OK (and expected in Colab) to use `!pip install ___` to install a library. In Colab, each notebook is independent of other notebooks, so the installation of packages has limited impact. But, be especially careful doing this on HPC systems; installing things that cause compatibility issues is easy.
>
>### For Jupyter on HPC systems
>
>For most of what the *Practicum AI* courses cover, an environment (known as a **kernel**) will be ready for you to use. Check the kernel if you run into an error about a missing library. The current kernel is shown in the top right of the notebook:
>![Screenshot of a Jupyter notebook showing the currently selected kernel](images/kernel.png)
>For this notebook on HiPerGator, we will use the **UFRC Python-3.10** kernel.

## pandas Dataframes

pandas provides one of the most useful data structures for managing data: the DataFrame.

**DataFrames:**

* Allow fast, flexible, and efficient data manipulation.
* Are the natural data structure for **tabular** data.
* Are two-dimensional, containing both **rows** and **columns**.
* Consist of one or more **Series** (as columns).


***pandas*** is built on top of the **NumPy** library, which in practice means that most of the methods defined for NumPy Arrays apply to pandas Series/DataFrames.

**What makes pandas so useful for working with data?**
1. Powerful and efficient interface to access and manipulate individual records.
2. Tools for handling missing values and explicit values for missing data (`NaN` and `None`).
3. Relational database-style operations between DataFrames.

### <img src='images/note_icon.svg' alt="Note icon" width=40 align=center> Note
> Some of the examples used in this notebook are different from those used in the corresponding video.

## Creating a Series

Let's start by creating our first pandas Series! (Remember: in the context of a **DataFrame** (table), each column is represented by a **Series**).

First, let's create some example data to illustrate the creation of our Series. in this example, we will use data that relates a person's "SpO2" reading from a pulse oximeter (which shows the percentage of oxygen in someone's blood), and their risk of adverse outcomes.

In [ ]:
spo2 = [88, 99, 88, 85, 100, 98] # Review: this is a Python list, containing 6 values.
risk = ['high', 'low', 'high', 'high', 'low', 'low']

In [ ]:
# Create our first Series.
spo2_series = pd.Series(spo2)
spo2_series # If a variable name is the last line in a code cell, Jupyter will display the contents of the variable.

0     88
1     99
2     88
3     85
4    100
5     98
dtype: int64

In [ ]:
# Do the same for risk.
risk_series = pd.Series(risk) # Display the 'risk' Series

#### I can also a series of two list

In [ ]:
spo_risk = pd.Series(data=spo2, index=risk)
spo_risk

high     88
low      99
high     88
high     85
low     100
low      98
dtype: int64

#### Or in this way

In [ ]:
spo_risk2 = pd.Series(spo2, risk)
spo_risk2

high     88
low      99
high     88
high     85
low     100
low      98
dtype: int64

### <img src='images/alert_icon.svg' alt="Alert icon" width=40 align=center> Alert!
> An important point to remember is that Python is "zero-indexed." That means that the first item in a list, or the first row in a table, has the index of 0.
>
> Notice the output above. The first column shows the Series index, and the second column has the value. The indices start at 0!
>
> This applies to lists, strings, arrays, and almost everything in Python. Keep this in mind--a common error is mistakenly using index 1, thinking you are getting the first item, where you need to use index 0.


## Creating a pandas Dataframe

* Now that we've seen a Series, we can see how these are combined as **columns** in a **DataFrame**.
* By popular convention, we will name our DataFrame variable `df`.

### <img src='images/example_icon.svg' alt="Example icon" width=40 align=center> Example

*  Create a small DataFrame from our example data
*  You need to write spo2 and risk twice when creating the DataFrame because you are defining the column names and their corresponding data. The first instance of spo2 and risk specifies the column names, and the second instance provides the actual data for those columns.



In [ ]:
df = pd.DataFrame({'spo2': spo2, 'risk': risk}) # Create the DataFrame
df # View the DataFrame

spo2  risk
0    88  high
1    99   low
2    88  high
3    85  high
4   100   low
5    98   low

### <img src='images/example_icon.svg' alt="Example icon" width=40 align=center> Example
> Create a larger DataFrame from random data.

In [ ]:
# Make some random numbers for data.
# This makes an array of random integers from 1-10, then reshapes it into a
# 500x4 matrix (500 rows and 4 columns).
random_data = np.random.randint(1,10, size=2000).reshape(500, 4)

# Create some example column names
column_names = ['columnA', 'columnB', 'columnC', 'columnD']

# Create the DateFrame from the random data we generated.
# Since the input matrix has 500 rows and 4 columns, the DataFrame will have
# 500 rows and 4 columns.
df = pd.DataFrame(data=random_data, columns=column_names)
df # View the DataFrame.

columnA  columnB  columnC  columnD
0          2        8        4        1
1          2        6        3        4
2          3        2        7        4
3          5        7        8        2
4          3        9        5        6
..       ...      ...      ...      ...
495        4        7        7        1
496        1        8        2        3
497        9        8        9        8
498        9        7        7        1
499        7        5        1        6

[500 rows x 4 columns]

### <img src='images/note_icon.svg' alt="Note icon" width=40 align=center> Note
> When viewing a large DataFrame in a Jupyter notebook, sometimes the data will be **truncated** as above.

We often want to see the first (or last) several rows of a DateFrame. We can use the `head()` and `tail()` functions.

In [ ]:
# View the 1st 5 rows.
df.head()

columnA  columnB  columnC  columnD
0        2        8        4        1
1        2        6        3        4
2        3        2        7        4
3        5        7        8        2
4        3        9        5        6

In [ ]:
# View the last 3 rows (head also takes a number of rows as an optional
# argument).
df.tail(3)

columnA  columnB  columnC  columnD
497        9        8        9        8
498        9        7        7        1
499        7        5        1        6

We can also limit our views to particular columns using the column names. If the column names have no spaces or special characters, we can use the "dot" notation: `df.columnA`. Otherwise, we must pass a list of columns: `df[['columnA', 'columnB']]`.

In [ ]:
# View multiple columns or if there are spaces in names
# Note the double square brackets!
df[['columnA', 'columnB']].head()

columnA  columnB
0        2        8
1        2        6
2        3        2
3        5        7
4        3        9

# Creating a DataFrame from an existing data file

* Instead of creating a DataFrame from scratch (tedious!), we usually create a DataFrame by loading an existing data file.
* The most common file format for storing datasets is the **comma-separated value** `(.csv)` file.

> To create a pandas DataFrame from a `.csv` file, we'll use the `read_csv` function.

### Let's load some data

The main input argument to the `read_csv` function can be a local filename or a remote file URL. In this example, we'll be using an online file.

The dataset used is from the [GapMinder Project](https://www.gapminder.org/) and includes information on GDP, life expectancy, and population in various countries for different years.

In [ ]:
url='https://raw.githubusercontent.com/PracticumAI/python-for-ai/main/data/gapminder_all.csv' # Direct link to the dataset on GitHub
df = pd.read_csv(url) # Create the DataFrame from the given URL
df # View the DateFrame we just created

continent         country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0      Africa         Algeria     2449.008185     3013.976023     2550.816880   
1      Africa          Angola     3520.610273     3827.940465     4269.276742   
2      Africa           Benin     1062.752200      959.601080      949.499064   
3      Africa        Botswana      851.241141      918.232535      983.653976   
4      Africa    Burkina Faso      543.255241      617.183465      722.512021   
..        ...             ...             ...             ...             ...   
137    Europe     Switzerland    14734.232750    17909.489730    20431.092700   
138    Europe          Turkey     1969.100980     2218.754257     2322.869908   
139    Europe  United Kingdom     9979.508487    11283.177950    12477.177070   
140   Oceania       Australia    10039.595640    10949.649590    12217.226860   
141   Oceania     New Zealand    10556.575660    12247.395320    13175.678000   

     gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0       3246.991771     4182.663766     4910.416756     5745.160213   
1       5522.776375     5473.288005     3008.647355     2756.953672   
2       1035.831411     1085.796879     1029.161251     1277.897616   
3       1214.709294     2263.611114     3214.857818     4551.142150   
4        794.826560      854.735976      743.387037      807.198586   
..              ...             ...             ...             ...   
137    22966.144320    27195.113040    26982.290520    28397.715120   
138     2826.356387     3450.696380     4269.122326     4241.356344   
139    14142.850890    15895.116410    17428.748460    18232.424520   
140    14526.124650    16788.629480    18334.197510    19477.009280   
141    14463.918930    16046.037280    16233.717700    17632.410400   

     gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972    pop_1977  \
0       5681.358539  ...  11000948.0  12760499.0  14760787.0  17152804.0   
1       2430.208311  ...   4826015.0   5247469.0   5894858.0   6162675.0   
2       1225.856010  ...   2151895.0   2427334.0   2761407.0   3168267.0   
3       6205.883850  ...    512764.0    553541.0    619351.0    781472.0   
4        912.063142  ...   4919632.0   5127935.0   5433886.0   5889574.0   
..              ...  ...         ...         ...         ...         ...   
137    30281.704590  ...   5666000.0   6063000.0   6401400.0   6316424.0   
138     5089.043686  ...  29788695.0  33411317.0  37492953.0  42404033.0   
139    21664.787670  ...  53292000.0  54959000.0  56079000.0  56179000.0   
140    21888.889030  ...  10794968.0  11872264.0  13177000.0  14074100.0   
141    19007.191290  ...   2488550.0   2728150.0   2929100.0   3164900.0   

       pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  
0    20033753.0  23254956.0  26298373.0  29072015.0  31287142  33333216  
1     7016384.0   7874230.0   8735988.0   9875024.0  10866106  12420476  
2     3641603.0   4243788.0   4981671.0   6066080.0   7026113   8078314  
3      970347.0   1151184.0   1342614.0   1536536.0   1630347   1639131  
4     6634596.0   7586551.0   8878303.0  10352843.0  12251209  14326203  
..          ...         ...         ...         ...       ...       ...  
137   6468126.0   6649942.0   6995447.0   7193761.0   7361757   7554661  
138  47328791.0  52881328.0  58179144.0  63047647.0  67308928  71158647  
139  56339704.0  56981620.0  57866349.0  58808266.0  59912431  60776238  
140  15184200.0  16257249.0  17481977.0  18565243.0  19546792  20434176  
141   3210650.0   3317166.0   3437674.0   3676187.0   3908037   4115771  

[142 rows x 38 columns]

We can see that this dataset contains 142 rows and 38 columns corresponding to each country's GDP, life expectancy, and population between 1952 and 2007.

### <img src='images/note_icon.svg' alt="Note icon" width=40 align=center> Note
> Stop for a second here. What happened to the variable `df` we defined for our SpO2 data?


## Viewing parts of a Dataframe
* Dataframes often hold many rows (in many AI projects, you might be dealing with hundreds of thousands, or even millions, of samples!)
* Sometimes, we want a preview or a sense of the data stored in our DataFrame.
* In the below examples, we'll see a few ways to view parts of a DataFrame.
    * This includes viewing a subset of columns, a subset of rows, or both.

With so many columns, it can be a bit hard to see all of them, but we can get a list of the columns in the DataFrame with the `.columns` method:

In [ ]:
df.columns

Index(['continent', 'country', 'gdpPercap_1952', 'gdpPercap_1957',
       'gdpPercap_1962', 'gdpPercap_1967', 'gdpPercap_1972', 'gdpPercap_1977',
       'gdpPercap_1982', 'gdpPercap_1987', 'gdpPercap_1992', 'gdpPercap_1997',
       'gdpPercap_2002', 'gdpPercap_2007', 'lifeExp_1952', 'lifeExp_1957',
       'lifeExp_1962', 'lifeExp_1967', 'lifeExp_1972', 'lifeExp_1977',
       'lifeExp_1982', 'lifeExp_1987', 'lifeExp_1992', 'lifeExp_1997',
       'lifeExp_2002', 'lifeExp_2007', 'pop_1952', 'pop_1957', 'pop_1962',
       'pop_1967', 'pop_1972', 'pop_1977', 'pop_1982', 'pop_1987', 'pop_1992',
       'pop_1997', 'pop_2002', 'pop_2007'],
      dtype='object')

Next, let's revisit viewing some of a DataFrame's **rows**. Since pandas is so flexible, there are several ways to accomplish this. The examples below use the pandas' functions `head` and `tail`.

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 1

> Use `head` to view the first 5 rows of our `df` DataFrame. (**Hint:** Given a DataFrame `X`, calling `X.head()` will show the first `5` rows of the DataFrame.  

In [ ]:
# View only the first 5 rows of the DateFrame.
df.head(5)

continent       country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0    Africa       Algeria     2449.008185     3013.976023     2550.816880   
1    Africa        Angola     3520.610273     3827.940465     4269.276742   
2    Africa         Benin     1062.752200      959.601080      949.499064   
3    Africa      Botswana      851.241141      918.232535      983.653976   
4    Africa  Burkina Faso      543.255241      617.183465      722.512021   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     3246.991771     4182.663766     4910.416756     5745.160213   
1     5522.776375     5473.288005     3008.647355     2756.953672   
2     1035.831411     1085.796879     1029.161251     1277.897616   
3     1214.709294     2263.611114     3214.857818     4551.142150   
4      794.826560      854.735976      743.387037      807.198586   

   gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972    pop_1977  \
0     5681.358539  ...  11000948.0  12760499.0  14760787.0  17152804.0   
1     2430.208311  ...   4826015.0   5247469.0   5894858.0   6162675.0   
2     1225.856010  ...   2151895.0   2427334.0   2761407.0   3168267.0   
3     6205.883850  ...    512764.0    553541.0    619351.0    781472.0   
4      912.063142  ...   4919632.0   5127935.0   5433886.0   5889574.0   

     pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  
0  20033753.0  23254956.0  26298373.0  29072015.0  31287142  33333216  
1   7016384.0   7874230.0   8735988.0   9875024.0  10866106  12420476  
2   3641603.0   4243788.0   4981671.0   6066080.0   7026113   8078314  
3    970347.0   1151184.0   1342614.0   1536536.0   1630347   1639131  
4   6634596.0   7586551.0   8878303.0  10352843.0  12251209  14326203  

[5 rows x 38 columns]

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 2
> Use `head` to view the first **10** rows of `df`. (**Hint:** You can also provide an input argument `n` to the `head` function to show the first `n` rows (instead of the default `5` rows)).

In [ ]:
# View only the first 10 rows of the DataFrame.
df.head(10)

continent                   country  gdpPercap_1952  gdpPercap_1957  \
0    Africa                   Algeria     2449.008185     3013.976023   
1    Africa                    Angola     3520.610273     3827.940465   
2    Africa                     Benin     1062.752200      959.601080   
3    Africa                  Botswana      851.241141      918.232535   
4    Africa              Burkina Faso      543.255241      617.183465   
5    Africa                   Burundi      339.296459      379.564628   
6    Africa                  Cameroon     1172.667655     1313.048099   
7    Africa  Central African Republic     1071.310713     1190.844328   
8    Africa                      Chad     1178.665927     1308.495577   
9    Africa                   Comoros     1102.990936     1211.148548   

   gdpPercap_1962  gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
0     2550.816880     3246.991771     4182.663766     4910.416756   
1     4269.276742     5522.776375     5473.288005     3008.647355   
2      949.499064     1035.831411     1085.796879     1029.161251   
3      983.653976     1214.709294     2263.611114     3214.857818   
4      722.512021      794.826560      854.735976      743.387037   
5      355.203227      412.977514      464.099504      556.103265   
6     1399.607441     1508.453148     1684.146528     1783.432873   
7     1193.068753     1136.056615     1070.013275     1109.374338   
8     1389.817618     1196.810565     1104.103987     1133.984950   
9     1406.648278     1876.029643     1937.577675     1172.603047   

   gdpPercap_1982  gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972  \
0     5745.160213     5681.358539  ...  11000948.0  12760499.0  14760787.0   
1     2756.953672     2430.208311  ...   4826015.0   5247469.0   5894858.0   
2     1277.897616     1225.856010  ...   2151895.0   2427334.0   2761407.0   
3     4551.142150     6205.883850  ...    512764.0    553541.0    619351.0   
4      807.198586      912.063142  ...   4919632.0   5127935.0   5433886.0   
5      559.603231      621.818819  ...   2961915.0   3330989.0   3529983.0   
6     2367.983282     2602.664206  ...   5793633.0   6335506.0   7021028.0   
7      956.752991      844.876350  ...   1523478.0   1733638.0   1927260.0   
8      797.908101      952.386129  ...   3150417.0   3495967.0   3899068.0   
9     1267.100083     1315.980812  ...    191689.0    217378.0    250027.0   

     pop_1977    pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  \
0  17152804.0  20033753.0  23254956.0  26298373.0  29072015.0  31287142   
1   6162675.0   7016384.0   7874230.0   8735988.0   9875024.0  10866106   
2   3168267.0   3641603.0   4243788.0   4981671.0   6066080.0   7026113   
3    781472.0    970347.0   1151184.0   1342614.0   1536536.0   1630347   
4   5889574.0   6634596.0   7586551.0   8878303.0  10352843.0  12251209   
5   3834415.0   4580410.0   5126023.0   5809236.0   6121610.0   7021078   
6   7959865.0   9250831.0  10780667.0  12467171.0  14195809.0  15929988   
7   2167533.0   2476971.0   2840009.0   3265124.0   3696513.0   4048013   
8   4388260.0   4875118.0   5498955.0   6429417.0   7562011.0   8835739   
9    304739.0    348643.0    395114.0    454429.0    527982.0    614382   

   pop_2007  
0  33333216  
1  12420476  
2   8078314  
3   1639131  
4  14326203  
5   8390505  
6  17696293  
7   4369038  
8  10238807  
9    710960  

[10 rows x 38 columns]

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 3
> Use `tail` to view the **last** 5 rows of our `df` DataFrame. (**Hint:** Given a DateFrame `X`, calling `X.tail()` will show the **last** `5` rows of the DateFrame).  

In [ ]:
# View the last 5 rows of the DataFrame.
df.tail(5)

continent         country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
137    Europe     Switzerland    14734.232750    17909.489730    20431.092700   
138    Europe          Turkey     1969.100980     2218.754257     2322.869908   
139    Europe  United Kingdom     9979.508487    11283.177950    12477.177070   
140   Oceania       Australia    10039.595640    10949.649590    12217.226860   
141   Oceania     New Zealand    10556.575660    12247.395320    13175.678000   

     gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
137    22966.144320     27195.11304    26982.290520    28397.715120   
138     2826.356387      3450.69638     4269.122326     4241.356344   
139    14142.850890     15895.11641    17428.748460    18232.424520   
140    14526.124650     16788.62948    18334.197510    19477.009280   
141    14463.918930     16046.03728    16233.717700    17632.410400   

     gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972    pop_1977  \
137    30281.704590  ...   5666000.0   6063000.0   6401400.0   6316424.0   
138     5089.043686  ...  29788695.0  33411317.0  37492953.0  42404033.0   
139    21664.787670  ...  53292000.0  54959000.0  56079000.0  56179000.0   
140    21888.889030  ...  10794968.0  11872264.0  13177000.0  14074100.0   
141    19007.191290  ...   2488550.0   2728150.0   2929100.0   3164900.0   

       pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  
137   6468126.0   6649942.0   6995447.0   7193761.0   7361757   7554661  
138  47328791.0  52881328.0  58179144.0  63047647.0  67308928  71158647  
139  56339704.0  56981620.0  57866349.0  58808266.0  59912431  60776238  
140  15184200.0  16257249.0  17481977.0  18565243.0  19546792  20434176  
141   3210650.0   3317166.0   3437674.0   3676187.0   3908037   4115771  

[5 rows x 38 columns]

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 4
> Use `tail` to view the **last** `10` rows of `df`. (**Hint:** You can also provide an input argument `n` to the `tail` function to show the last `n` rows (instead of the default `5` rows)).

In [ ]:
# View only the last 10 rows of the DateFrame.
df.tail(10)

continent          country  gdpPercap_1952  gdpPercap_1957  \
132    Europe           Serbia     3581.459448     4981.090891   
133    Europe  Slovak Republic     5074.659104     6093.262980   
134    Europe         Slovenia     4215.041741     5862.276629   
135    Europe            Spain     3834.034742     4564.802410   
136    Europe           Sweden     8527.844662     9911.878226   
137    Europe      Switzerland    14734.232750    17909.489730   
138    Europe           Turkey     1969.100980     2218.754257   
139    Europe   United Kingdom     9979.508487    11283.177950   
140   Oceania        Australia    10039.595640    10949.649590   
141   Oceania      New Zealand    10556.575660    12247.395320   

     gdpPercap_1962  gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
132     6289.629157     7991.707066    10522.067490    12980.669560   
133     7481.107598     8412.902397     9674.167626    10922.664040   
134     7402.303395     9405.489397    12383.486200    15277.030170   
135     5693.843879     7993.512294    10638.751310    13236.921170   
136    12329.441920    15258.296970    17832.024640    18855.725210   
137    20431.092700    22966.144320    27195.113040    26982.290520   
138     2322.869908     2826.356387     3450.696380     4269.122326   
139    12477.177070    14142.850890    15895.116410    17428.748460   
140    12217.226860    14526.124650    16788.629480    18334.197510   
141    13175.678000    14463.918930    16046.037280    16233.717700   

     gdpPercap_1982  gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972  \
132    15181.092700    15870.878510  ...   7616060.0   7971222.0   8313288.0   
133    11348.545850    12037.267580  ...   4237384.0   4442238.0   4593433.0   
134    17866.721750    18678.534920  ...   1582962.0   1646912.0   1694510.0   
135    13926.169970    15764.983130  ...  31158061.0  32850275.0  34513161.0   
136    20667.381250    23586.929270  ...   7561588.0   7867931.0   8122293.0   
137    28397.715120    30281.704590  ...   5666000.0   6063000.0   6401400.0   
138     4241.356344     5089.043686  ...  29788695.0  33411317.0  37492953.0   
139    18232.424520    21664.787670  ...  53292000.0  54959000.0  56079000.0   
140    19477.009280    21888.889030  ...  10794968.0  11872264.0  13177000.0   
141    17632.410400    19007.191290  ...   2488550.0   2728150.0   2929100.0   

       pop_1977    pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  \
132   8686367.0   9032824.0   9230783.0   9826397.0  10336594.0  10111559   
133   4827803.0   5048043.0   5199318.0   5302888.0   5383010.0   5410052   
134   1746919.0   1861252.0   1945870.0   1999210.0   2011612.0   2011497   
135  36439000.0  37983310.0  38880702.0  39549438.0  39855442.0  40152517   
136   8251648.0   8325260.0   8421403.0   8718867.0   8897619.0   8954175   
137   6316424.0   6468126.0   6649942.0   6995447.0   7193761.0   7361757   
138  42404033.0  47328791.0  52881328.0  58179144.0  63047647.0  67308928   
139  56179000.0  56339704.0  56981620.0  57866349.0  58808266.0  59912431   
140  14074100.0  15184200.0  16257249.0  17481977.0  18565243.0  19546792   
141   3164900.0   3210650.0   3317166.0   3437674.0   3676187.0   3908037   

     pop_2007  
132  10150265  
133   5447502  
134   2009245  
135  40448191  
136   9031088  
137   7554661  
138  71158647  
139  60776238  
140  20434176  
141   4115771  

[10 rows x 38 columns]

Now, let's see some ways to view a subset of a DateFrame's **columns**.

**Reminder:** If the last line inside a code cell is simply a variable name, Jupyter will display the value of that variable.

For a given DateFrame `df` and desired column `col`, one way to view a single column is to write `df['col']`. (**Important:** the column name `col` must be in quotes!).

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 5
> Use the above method to view the `gdpPercap_2007` column of our `df` DateFrame.

In [ ]:
# View the gdpPercap_2007 column of the DateFrame.
df['gdpPercap_2007']

0       6223.367465
1       4797.231267
2       1441.284873
3      12569.851770
4       1217.032994
           ...     
137    37506.419070
138     8458.276384
139    33203.261280
140    34435.367440
141    25185.009110
Name: gdpPercap_2007, Length: 142, dtype: float64

We can also adapt this method to view multiple columns. Instead of passing in a single string column name `'C'`, we will pass in a **list** of string column names, e.g. `['A', 'B', 'C']`. For example: `df[['A', 'B', 'C']]`

### <img src='images/note_icon.svg' alt="Note icon" width=40 align=center> Note
> **Important:** Notice the double brackets! DataFrames use a similar syntax as lists. To get a single column, we can use the column's name: e.g., `df['A']` would give us column A. Here, we want multiple columns, so a list of column names is provided.

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 6
> Use the above method to view the `gdpPercap_2007`, `lifeExp_2007`, and `pop_2007` columns of our `df` DateFrame.

In [ ]:
# Display the columns `gdpPercap_2007`, `lifeExp_2007`, and `pop_2007`
df[['gdpPercap_2007', 'lifeExp_2007', 'pop_2007']]


gdpPercap_2007  lifeExp_2007  pop_2007
0       6223.367465        72.301  33333216
1       4797.231267        42.731  12420476
2       1441.284873        56.728   8078314
3      12569.851770        50.728   1639131
4       1217.032994        52.295  14326203
..              ...           ...       ...
137    37506.419070        81.701   7554661
138     8458.276384        71.777  71158647
139    33203.261280        79.425  60776238
140    34435.367440        81.235  20434176
141    25185.009110        80.204   4115771

[142 rows x 3 columns]

## Creating new columns

* We can easily create new columns in an existing DateFrame.
* When we use mathematical operations as below, they will be applied to every element of the Series.
* This way, we can create new columns from different combinations of existing ones.

Let's use an equation to calculate the change in GDP from 1962 to 2007.

In [ ]:
df['gdp_change'] = (df['gdpPercap_1962'] - df['gdpPercap_2007'])
df.head()

continent       country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0    Africa       Algeria     2449.008185     3013.976023     2550.816880   
1    Africa        Angola     3520.610273     3827.940465     4269.276742   
2    Africa         Benin     1062.752200      959.601080      949.499064   
3    Africa      Botswana      851.241141      918.232535      983.653976   
4    Africa  Burkina Faso      543.255241      617.183465      722.512021   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     3246.991771     4182.663766     4910.416756     5745.160213   
1     5522.776375     5473.288005     3008.647355     2756.953672   
2     1035.831411     1085.796879     1029.161251     1277.897616   
3     1214.709294     2263.611114     3214.857818     4551.142150   
4      794.826560      854.735976      743.387037      807.198586   

   gdpPercap_1987  ...    pop_1967    pop_1972    pop_1977    pop_1982  \
0     5681.358539  ...  12760499.0  14760787.0  17152804.0  20033753.0   
1     2430.208311  ...   5247469.0   5894858.0   6162675.0   7016384.0   
2     1225.856010  ...   2427334.0   2761407.0   3168267.0   3641603.0   
3     6205.883850  ...    553541.0    619351.0    781472.0    970347.0   
4      912.063142  ...   5127935.0   5433886.0   5889574.0   6634596.0   

     pop_1987    pop_1992    pop_1997  pop_2002  pop_2007    gdp_change  
0  23254956.0  26298373.0  29072015.0  31287142  33333216  -3672.550585  
1   7874230.0   8735988.0   9875024.0  10866106  12420476   -527.954525  
2   4243788.0   4981671.0   6066080.0   7026113   8078314   -491.785809  
3   1151184.0   1342614.0   1536536.0   1630347   1639131 -11586.197794  
4   7586551.0   8878303.0  10352843.0  12251209  14326203   -494.520973  

[5 rows x 39 columns]

## Deleting columns

We can also **drop** (delete) columns.

Let's remove the `gdp_change` column that we previously created.

In [ ]:
df.drop(columns=['gdp_change']) # Will this remove the column?
df.head()

continent       country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0    Africa       Algeria     2449.008185     3013.976023     2550.816880   
1    Africa        Angola     3520.610273     3827.940465     4269.276742   
2    Africa         Benin     1062.752200      959.601080      949.499064   
3    Africa      Botswana      851.241141      918.232535      983.653976   
4    Africa  Burkina Faso      543.255241      617.183465      722.512021   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     3246.991771     4182.663766     4910.416756     5745.160213   
1     5522.776375     5473.288005     3008.647355     2756.953672   
2     1035.831411     1085.796879     1029.161251     1277.897616   
3     1214.709294     2263.611114     3214.857818     4551.142150   
4      794.826560      854.735976      743.387037      807.198586   

   gdpPercap_1987  ...    pop_1967    pop_1972    pop_1977    pop_1982  \
0     5681.358539  ...  12760499.0  14760787.0  17152804.0  20033753.0   
1     2430.208311  ...   5247469.0   5894858.0   6162675.0   7016384.0   
2     1225.856010  ...   2427334.0   2761407.0   3168267.0   3641603.0   
3     6205.883850  ...    553541.0    619351.0    781472.0    970347.0   
4      912.063142  ...   5127935.0   5433886.0   5889574.0   6634596.0   

     pop_1987    pop_1992    pop_1997  pop_2002  pop_2007    gdp_change  
0  23254956.0  26298373.0  29072015.0  31287142  33333216  -3672.550585  
1   7874230.0   8735988.0   9875024.0  10866106  12420476   -527.954525  
2   4243788.0   4981671.0   6066080.0   7026113   8078314   -491.785809  
3   1151184.0   1342614.0   1536536.0   1630347   1639131 -11586.197794  
4   7586551.0   8878303.0  10352843.0  12251209  14326203   -494.520973  

[5 rows x 39 columns]

**What happened? `gdp_change` is still there!**

### <img src='images/tip_icon.svg' alt="Tip icon" width=40 align=center> Tip
> Unless you specify `inplace=True` or assign the DateFrame back to itself (or a different variable name), **the `drop` method does not change the original DateFrame!**

Below is one way to permanently remove a column from a DateFrame by assigning the result to a new DateFrame variable `df2`.

In [ ]:
# After dropping the column, we assign the resulting DateFrame to a new variable.
# This will persist the change.
df2 = df.drop(columns=['gdp_change'])
df2.head()

continent       country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0    Africa       Algeria     2449.008185     3013.976023     2550.816880   
1    Africa        Angola     3520.610273     3827.940465     4269.276742   
2    Africa         Benin     1062.752200      959.601080      949.499064   
3    Africa      Botswana      851.241141      918.232535      983.653976   
4    Africa  Burkina Faso      543.255241      617.183465      722.512021   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     3246.991771     4182.663766     4910.416756     5745.160213   
1     5522.776375     5473.288005     3008.647355     2756.953672   
2     1035.831411     1085.796879     1029.161251     1277.897616   
3     1214.709294     2263.611114     3214.857818     4551.142150   
4      794.826560      854.735976      743.387037      807.198586   

   gdpPercap_1987  ...    pop_1962    pop_1967    pop_1972    pop_1977  \
0     5681.358539  ...  11000948.0  12760499.0  14760787.0  17152804.0   
1     2430.208311  ...   4826015.0   5247469.0   5894858.0   6162675.0   
2     1225.856010  ...   2151895.0   2427334.0   2761407.0   3168267.0   
3     6205.883850  ...    512764.0    553541.0    619351.0    781472.0   
4      912.063142  ...   4919632.0   5127935.0   5433886.0   5889574.0   

     pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  
0  20033753.0  23254956.0  26298373.0  29072015.0  31287142  33333216  
1   7016384.0   7874230.0   8735988.0   9875024.0  10866106  12420476  
2   3641603.0   4243788.0   4981671.0   6066080.0   7026113   8078314  
3    970347.0   1151184.0   1342614.0   1536536.0   1630347   1639131  
4   6634596.0   7586551.0   8878303.0  10352843.0  12251209  14326203  

[5 rows x 38 columns]

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 7

> Create a new DateFrame called `df3`, that contains everything in the `df` DateFrame **except** the `continent` column.

In [ ]:
df3 = df.drop(columns=['continent'])
df3.head()


country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0       Algeria     2449.008185     3013.976023     2550.816880   
1        Angola     3520.610273     3827.940465     4269.276742   
2         Benin     1062.752200      959.601080      949.499064   
3      Botswana      851.241141      918.232535      983.653976   
4  Burkina Faso      543.255241      617.183465      722.512021   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     3246.991771     4182.663766     4910.416756     5745.160213   
1     5522.776375     5473.288005     3008.647355     2756.953672   
2     1035.831411     1085.796879     1029.161251     1277.897616   
3     1214.709294     2263.611114     3214.857818     4551.142150   
4      794.826560      854.735976      743.387037      807.198586   

   gdpPercap_1987  gdpPercap_1992  ...    pop_1967    pop_1972    pop_1977  \
0     5681.358539     5023.216647  ...  12760499.0  14760787.0  17152804.0   
1     2430.208311     2627.845685  ...   5247469.0   5894858.0   6162675.0   
2     1225.856010     1191.207681  ...   2427334.0   2761407.0   3168267.0   
3     6205.883850     7954.111645  ...    553541.0    619351.0    781472.0   
4      912.063142      931.752773  ...   5127935.0   5433886.0   5889574.0   

     pop_1982    pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  \
0  20033753.0  23254956.0  26298373.0  29072015.0  31287142  33333216   
1   7016384.0   7874230.0   8735988.0   9875024.0  10866106  12420476   
2   3641603.0   4243788.0   4981671.0   6066080.0   7026113   8078314   
3    970347.0   1151184.0   1342614.0   1536536.0   1630347   1639131   
4   6634596.0   7586551.0   8878303.0  10352843.0  12251209  14326203   

     gdp_change  
0  -3672.550585  
1   -527.954525  
2   -491.785809  
3 -11586.197794  
4   -494.520973  

[5 rows x 38 columns]

## Descriptive Statistics

### <img src='images/alert_icon.svg' alt="Alert icon" width=40 align=center> Alert!
> Before diving into an AI project, becoming familiar with the data you will be modeling is critical!  pandas has an extensive collection of optimized methods designed to help you analyze your data.

A preliminary data discovery and exploration phase can help you better understand the data you will be modeling and has many downstream advantages:
* You may identify potential problems with the dataset (e.g., missing values, outliers, incomplete columns, irrelevant information).
* Results of data exploration may help you decide which preprocessing or modeling approach to take.
* A better understanding of your dataset will help you interpret future results.

#### Remember Python functions?
pandas offers many **functions** that take as **input** a set of values (example: all of the values from the `gdpPercap_2007` column) and **return** a calculated **output** value based on which function was used.

Here are a few examples (most are self-explanatory!):
* `sum` (the sum of the values)
* `count` (the number of values), `value_counts` (the number unique values)
* `min` (the minimum value), `max` (the maximum value)
* `mean` (the mean value), `median` (the median value), `mode` (the mode of the values)
* `var` (the variance of the values), `std` (the standard deviation of the values)
* `quantile` (the quantiles of the values)

Let's take a look at a few examples using our dataset!

### <img src='images/example_icon.svg' alt="Example icon" width=40 align=center> Example
Here's how we can compute the **mean** GDP in 2007:

In [ ]:
# Compute the mean GDP in 2007
df['gdpPercap_2007'].mean()

11680.071819878167

It's that easy! Let's compute a few more descriptive statistics:

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 8
> Compute the **minimum** GDP in 2007.

In [ ]:
# Compute the minimum GDP in 2007
df['gdpPercap_2007'].min()

277.5518587

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 9
> Compute the **maximum** GDP in 2007.

In [ ]:
# Compute the maximum GDP in 2007
df['gdpPercap_2007'].max()

49357.19017

Finally, the `describe` function can be used on an entire DateFrame to display several standard statistics computed for each column. (**Note:** Be aware of each column's values and data types, as sometimes the statistics will not make sense.)

In [ ]:
# (Almost) everything, (almost) everywhere, all at once
# Note: will only compute statistics for columns with numerical data.
df.describe()

gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
count      142.000000      142.000000      142.000000      142.000000   
mean      3725.276046     4299.408345     4725.812342     5483.653047   
std       9321.064786     9869.662202     8667.362525     8095.315431   
min        298.846212      335.997115      355.203227      349.000000   
25%        864.752389      930.540819     1059.149171     1151.245103   
50%       1968.528344     2173.220291     2335.439533     2678.334740   
75%       3913.492777     4876.356362     5709.381428     7075.932943   
max     108382.352900   113523.132900    95458.111760    80894.883260   

       gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
count      142.000000      142.000000      142.000000      142.000000   
mean      6770.082815     7313.166421     7518.901673     7900.920218   
std      10614.383403     8362.489150     7733.845006     8288.281304   
min        357.000000      371.000000      424.000000      385.000000   
25%       1257.193853     1357.257252     1363.338985     1327.469823   
50%       3339.129407     3798.609244     4216.228428     4280.300366   
75%       9508.839304    11204.102423    12347.953722    11994.052795   
max     109347.867000    59265.477140    33693.175250    31540.974800   

       gdpPercap_1992  gdpPercap_1997  ...      pop_1967      pop_1972  \
count      142.000000      142.000000  ...  1.420000e+02  1.420000e+02   
mean      8158.608521     9090.175363  ...  2.265830e+07  2.518998e+07   
std       9031.846080    10171.493263  ...  7.837548e+07  8.864682e+07   
min        347.000000      312.188423  ...  7.078700e+04  7.659500e+04   
25%       1270.660958     1366.837958  ...  2.034768e+06  2.351192e+06   
50%       4386.085502     4781.825478  ...  5.170176e+06  5.877996e+06   
75%      10684.351870    12022.867188  ...  1.261458e+07  1.467920e+07   
max      34932.919590    41283.164330  ...  7.545500e+08  8.620300e+08   

           pop_1977      pop_1982      pop_1987      pop_1992      pop_1997  \
count  1.420000e+02  1.420000e+02  1.420000e+02  1.420000e+02  1.420000e+02   
mean   2.767638e+07  3.020730e+07  3.303857e+07  3.599092e+07  3.883947e+07   
std    9.748109e+07  1.050986e+08  1.147562e+08  1.245026e+08  1.334174e+08   
min    8.679600e+04  9.859300e+04  1.108120e+05  1.259110e+05  1.456080e+05   
25%    2.759717e+06  3.006286e+06  3.194990e+06  3.605992e+06  3.770150e+06   
50%    6.404036e+06  7.007320e+06  7.774862e+06  8.688686e+06  9.735064e+06   
75%    1.667023e+07  1.840732e+07  2.094754e+07  2.270538e+07  2.431137e+07   
max    9.434550e+08  1.000281e+09  1.084035e+09  1.164970e+09  1.230075e+09   

           pop_2002      pop_2007    gdp_change  
count  1.420000e+02  1.420000e+02    142.000000  
mean   4.145759e+07  4.402122e+07  -6954.259478  
std    1.408483e+08  1.476214e+08  10517.680561  
min    1.703720e+05  1.995790e+05 -43468.444068  
25%    4.173506e+06  4.508034e+06 -11012.551205  
50%    1.037292e+07  1.051753e+07  -2944.227486  
75%    2.654556e+07  3.121004e+07   -502.611723  
max    1.280400e+09  1.318683e+09  48151.121980  

[8 rows x 37 columns]

## Filtering data
pandas makes it easy to select, analyze, and process subsets of your data based on some criteria, `C`. This is accomplished with the syntax: `df[C]`, which will return a subset of DateFrame `df` that **matches** the criteria `C` listed inside brackets `[]`.

### Comparing values
In pandas, we can create filtering criteria using familiar mathematical operators to compare values:
* `>`: greater than
* `<`: less than
* `>=`: greater than or equal to
* `<=`: less than or equal to
* `!=`: is not equal to
* `==`: is equal to a single value
* **Bonus:** `.isin(L)`: equals any value in the list `L`

### <img src='images/example_icon.svg' alt="Example icon" width=40 align=center> Example
Display 2007 GDP is **greater** than $40,000.
* In this example, the criterion `C` is `df['gdpPercap_2007'] > 40000`.
* Recall the pandas filtering syntax: `df[C]`.
* So, this is what we want: `df[df['gdpPercap_2007'] > 40000]`.

In [ ]:
# Select rows where 2007 GDP is greater than $40,000
df[df['gdpPercap_2007'] > 40000]

continent        country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
74   Americas  United States    13990.482080    14847.127120    16173.145860   
92       Asia         Kuwait   108382.352900   113523.132900    95458.111760   
102      Asia      Singapore     2315.138227     2843.104409     3674.735572   
124    Europe        Ireland     5210.280328     5599.077872     6631.597314   
128    Europe         Norway    10095.421720    11653.973040    13450.401510   

     gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
74     19530.365570    21806.035940     24072.63213     25009.55914   
92     80894.883260   109347.867000     59265.47714     31354.03573   
102     4977.418540     8597.756202     11210.08948     15169.16112   
124     7655.568963     9530.772896     11150.98113     12618.32141   
128    16361.876470    18965.055510     23311.34939     26298.63531   

     gdpPercap_1987  ...     pop_1967     pop_1972     pop_1977     pop_1982  \
74      29884.35041  ...  198712000.0  209896000.0  220239000.0  232187835.0   
92      28118.42998  ...     575003.0     841934.0    1140357.0    1497494.0   
102     18861.53081  ...    1977600.0    2152400.0    2325300.0    2651869.0   
124     13872.86652  ...    2900100.0    3024400.0    3271900.0    3480000.0   
128     31540.97480  ...    3786019.0    3933004.0    4043205.0    4114787.0   

        pop_1987     pop_1992     pop_1997   pop_2002   pop_2007    gdp_change  
74   242803533.0  256894189.0  272911760.0  287675526  301139947 -26778.507230  
92     1891487.0    1418095.0    1765345.0    2111561    2505559  48151.121980  
102    2794552.0    3235865.0    3802309.0    4197776    4553009 -43468.444068  
124    3539900.0    3557761.0    3667233.0    3879155    4109086 -34044.399036  
128    4186147.0    4286357.0    4405672.0    4535591    4627926 -35906.788660  

[5 rows x 39 columns]

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 10
> Display where 2007 GDP is **less** than $500

In [ ]:
# Display rows where 2007 GDP is less than $500

In [ ]:
df[df['gdpPercap_2007'] < 500]

continent          country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
5     Africa          Burundi      339.296459      379.564628      355.203227   
10    Africa  Congo Dem. Rep.      780.542326      905.860230      896.314634   
25    Africa          Liberia      575.572996      620.969990      634.195163   
51    Africa         Zimbabwe      406.884115      518.764268      527.272182   

    gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
5       412.977514      464.099504      556.103265      559.603231   
10      861.593242      904.896068      795.757282      673.747818   
25      713.603648      803.005454      640.322438      572.199569   
51      569.795071      799.362176      685.587682      788.855041   

    gdpPercap_1987  ...    pop_1967    pop_1972    pop_1977    pop_1982  \
5       621.818819  ...   3330989.0   3529983.0   3834415.0   4580410.0   
10      672.774812  ...  19941073.0  23007669.0  26480870.0  30646495.0   
25      506.113857  ...   1279406.0   1482628.0   1703617.0   1956875.0   
51      706.157306  ...   4995432.0   5861135.0   6642107.0   7636524.0   

      pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  gdp_change  
5    5126023.0   5809236.0   6121610.0   7021078   8390505  -74.867464  
10  35481645.0  41672143.0  47798986.0  55379852  64606759  618.762775  
25   2269414.0   1912974.0   2200725.0   2814651   3193942  219.687821  
51   9216418.0  10704340.0  11404948.0  11926563  12311143   57.562884  

[4 rows x 39 columns]

### <img src='images/example_icon.svg' alt="Example icon" width=40 align=center> Example
Display the first 5 rows for African countries.

* Recall that the symbol `==` will check for equality (we want the continent to be equal to the single value `Africa`).
* In this example, the criterion `C` is `df['continent] == 'Africa'`.
* Given the pandas filtering syntax: `df[C]`, this is what we want: `df[df['continent'] == 'Africa']`.

In [ ]:
# Display the first 5 rows for African countries.
df[df['continent'] == 'Africa'].head()

continent       country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0    Africa       Algeria     2449.008185     3013.976023     2550.816880   
1    Africa        Angola     3520.610273     3827.940465     4269.276742   
2    Africa         Benin     1062.752200      959.601080      949.499064   
3    Africa      Botswana      851.241141      918.232535      983.653976   
4    Africa  Burkina Faso      543.255241      617.183465      722.512021   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     3246.991771     4182.663766     4910.416756     5745.160213   
1     5522.776375     5473.288005     3008.647355     2756.953672   
2     1035.831411     1085.796879     1029.161251     1277.897616   
3     1214.709294     2263.611114     3214.857818     4551.142150   
4      794.826560      854.735976      743.387037      807.198586   

   gdpPercap_1987  ...    pop_1967    pop_1972    pop_1977    pop_1982  \
0     5681.358539  ...  12760499.0  14760787.0  17152804.0  20033753.0   
1     2430.208311  ...   5247469.0   5894858.0   6162675.0   7016384.0   
2     1225.856010  ...   2427334.0   2761407.0   3168267.0   3641603.0   
3     6205.883850  ...    553541.0    619351.0    781472.0    970347.0   
4      912.063142  ...   5127935.0   5433886.0   5889574.0   6634596.0   

     pop_1987    pop_1992    pop_1997  pop_2002  pop_2007    gdp_change  
0  23254956.0  26298373.0  29072015.0  31287142  33333216  -3672.550585  
1   7874230.0   8735988.0   9875024.0  10866106  12420476   -527.954525  
2   4243788.0   4981671.0   6066080.0   7026113   8078314   -491.785809  
3   1151184.0   1342614.0   1536536.0   1630347   1639131 -11586.197794  
4   7586551.0   8878303.0  10352843.0  12251209  14326203   -494.520973  

[5 rows x 39 columns]

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 11
> Display the first 5 rows for European countries (`'continent'`==`'Europe'`).

In [ ]:
# Display the first 5 rows for European countries ('continent'=='Europe').
df[df['continent'] == 'Europe'].head()


continent                 country  gdpPercap_1952  gdpPercap_1957  \
110    Europe                 Albania     1601.056136     1942.284244   
111    Europe                 Austria     6137.076492     8842.598030   
112    Europe                 Belgium     8343.105127     9714.960623   
113    Europe  Bosnia and Herzegovina      973.533195     1353.989176   
114    Europe                Bulgaria     2444.286648     3008.670727   

     gdpPercap_1962  gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
110     2312.888958     2760.196931     3313.422188     3533.003910   
111    10750.721110    12834.602400    16661.625600    19749.422300   
112    10991.206760    13149.041190    16672.143560    19117.974480   
113     1709.683679     2172.352423     2860.169750     3528.481305   
114     4254.337839     5577.002800     6597.494398     7612.240438   

     gdpPercap_1982  gdpPercap_1987  ...   pop_1967   pop_1972   pop_1977  \
110     3630.880722     3738.932735  ...  1984060.0  2263554.0  2509048.0   
111    21597.083620    23687.826070  ...  7376998.0  7544201.0  7568430.0   
112    20979.845890    22525.563080  ...  9556500.0  9709100.0  9821800.0   
113     4126.613157     4314.114757  ...  3585000.0  3819000.0  4086000.0   
114     8224.191647     8239.854824  ...  8310226.0  8576200.0  8797022.0   

      pop_1982   pop_1987    pop_1992    pop_1997  pop_2002  pop_2007  \
110  2780097.0  3075321.0   3326498.0   3428038.0   3508512   3600523   
111  7574613.0  7578903.0   7914969.0   8069876.0   8148312   8199783   
112  9856303.0  9870200.0  10045622.0  10199787.0  10311970  10392226   
113  4172693.0  4338977.0   4256013.0   3607000.0   4165416   4552198   
114  8892098.0  8971958.0   8658506.0   8066057.0   7661799   7322858   

       gdp_change  
110  -3624.140568  
111 -25375.771590  
112 -22701.398320  
113  -5736.615124  
114  -6426.454981  

[5 rows x 39 columns]

### Combining multiple criteria

Multiple criteria can be combined using either the `&` or `|` symbols:
* C1 `&` C2: This will select rows that match **both** criterion C1 **and** criterion C2.
* C1 `|` C2: This will select rows that match **either** criterion C1 **or** criterion C2.

### <img src='images/example_icon.svg' alt="Example icon" width=40 align=center> Example
Display rows where 2007 GDP is greater than or equal to 3,000 and 2007 life expectancy is under 60. In this example, we have two criteria:
* C1: `df['gdpPercap_2007'] >= 3000`
* C2: `df['lifeExp_2007'] < 60`

In [ ]:
# Select rows where 2007 GDP is greater than or equal to 30,000 and 2007 life expectancy is under 60.
df[(df['gdpPercap_2007'] >= 3000) & (df['lifeExp_2007'] < 60)]

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 12
> Display all European countries that have 2007 life expectancies over 80.

In [1]:
# Display all European countries that have 2007 life expectancies over 80.
df[df['continent'] == 'Europe' & df['lifeExp_2007']>80]



NameError: name 'df' is not defined

### Analyzing filtered data
We can also perform our descriptive statistics on a filtered subset!

In [ ]:
df[(df['continent'] == 'Europe') & (df['lifeExp_2007'] > 80)].describe()

## Data visualization

Let's conclude this lesson by learning how pandas can enhance our understanding of a dataset with its many built-in visualization tools.

Let's see the data! This time, we'll load our dataset into a DateFrame variable named `X`.

> ![A hand icon](images/hand.png) Wait! `X`?? Didn't we say to avoid single-letter variable names and generally use lowercase?? Well...rules are...there to be broken when there is a good reason. In much of machine learning and AI, we work with datasets where we want to make a model to generate a prediction. Part of the goal is to take input data and predict an output or outcome. Thinking of a graph with `x`- and `y`-values, we can think of the `x`-value as an input and the `y`-value as the outcome. So, `x` and `y` make sense as variable names when thinking about values on a graph and datasets more generally. Going further, we borrow a convention from math, where a matrix (or a table with several columns) is referenced with a capital letter. Since our DataFrame has multiple columns, we will use the capital X for the name.
>
> We also want to ensure you understand that while we frequently use `df` for a DataFrame variable name, `df` is just a name, and nothing is special about it.

In [ ]:
X = pd.read_csv(url) # url is defined above
X.head()

### <img src='images/example_icon.svg' alt="Example icon" width=40 align=center> Example
Let's visually examine the distribution of 2007 GDPs with the pandas `hist` (histogram) function.

In [ ]:
hist = X['gdpPercap_2007'].hist()

We can modify many parameters of the `hist` function to create a more informative figure.

In [ ]:
hist = X['gdpPercap_2007'].hist(bins=72, grid=False)
hist.set_title('Distribution of 2007 GDP of Countries')
hist.set_ylabel('Count')
hist.set_xlabel('GDP in 2007')
display(hist)

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 13
> Visualize the distribution of the 2007 life expectancies.

In [ ]:
# Visualize the distribution of the 2007 life expectancies.
# Your code here


This can also be written as: `X.hist(['lifeExp_2007'], bins=40)`

That syntax makes it easier to plot multiple histograms side-by-side. Below, we compare the distributions of life expectancies in 1962 and 2007.

In [ ]:
X.hist(column=['lifeExp_1962','lifeExp_2007'], bins=40, sharey=True, sharex=True)

### <img src='images/example_icon.svg' alt="Example icon" width=40 align=center> Example
We can use the `df.plot.box` function to create box plots from columns in our DateFrame:

In [ ]:
X.plot.box(column='lifeExp_2007', by='continent')

### <img src='images/exercise_icon.svg' alt="Exercise icon" width=40 align=center> Exercise 14
> Practice with the `hist` and `box` functions by exploring different columns in our DateFrame. Experiment with data filtering methods to refine your figures further. **Hint:** We can see all available columns with `print(X.columns)`.

In [ ]:
### Experiment with code!


----
## Push changes to GitHub <img src="images/push_to_github.png" alt="Push to GitHub icon" align="right" width=150>

 Remember to **add**, **commit**, and **push** the changes you have made to this notebook to GitHub to keep your repository in sync.

In Jupyter, those are done in the git tab on the left. In Google Colab, use File > Save a copy in GitHub.


***
#### Attribution
Some content in this learning experience was adapted from Ben Shickel's [AI for Medicine bootcamp](https://github.com/gatorai/scripps) content.  